### ISSUE TITLE
Possible bug in RP conventions behavior (and ? on optional route params)

### ISSUE THREAD

**Issue by guardrex** [aspnet/Mvc#8385]
#### Is this a Bug or Feature request?:

Possible bug, but mostly in need of clarification for the docs. https://github.com/aspnet/Docs/issues/6173#issuecomment-386825256

#### Steps to reproduce:

App sets up a page route with an optional route param:

```csharp
options.Conventions.AddPageRoute("/Contact", "TheContactPage/{text?}");
```

#### Description of the problem:

Works when the param is present: `http://localhost:5000/TheContactPage/SomeValue`.

Doesn't work when the param isn't present (in spite of designation as optional): `http://localhost:5000/TheContactPage`. The app loads the app's Index page (i.e., no route matches).

Side-note: Upon rendering, the Contact page link in the Layout page points to `/TheContactPage` (`<a href="/TheContactPage">Contact</a>`) (good); but since that fails to resolve to the Contact page, it breaks the link.

The following is **not** a valid workaround ...

```csharp
options.Conventions.AddPageRoute("/Contact", "TheContactPage");
options.Conventions.AddPageRoute("/Contact", "TheContactPage/{text}");
```

The behavior doesn't change; and to make matters worse, the layout link to Contact (`<a asp-page="/Contact">Contact</a>`) isn't rendered to `/TheContactPage` ... it renders without a destination (`<a href="">Contact</a>`).

#### Version of `Microsoft.AspNetCore.Mvc` or `Microsoft.AspNetCore.App` or `Microsoft.AspNetCore.All`:

```xml
<PackageReference Include="Microsoft.AspNetCore.App" Version="2.1.3" />
```

#### Question

By-design? ... or bug? If by-design, I'll document that optional route params don't work with `AddPageRoute`.

**Comment by mkArtakMSFT** [comment]
Thanks @guardrex.
@rynowak, can you please answer this? Thanks!

**Comment by rynowak** [comment]
This all sounds really bizzare to me. I'll have to try and reproduce this to see what's going on. It's definitely intended that you can use optional parameters in pages routing.

@guardrex are there other pages at play here? Is there a page defined with the name `TheContactPage.cshtml`?

**Comment by guardrex** [comment]
Let me check and confirm that the topic sample is the exact repro for these behaviors. I think it is. brb

**Comment by rynowak** [comment]
Ya, if you have a link to an existing repro I'll just use that to investigate

**Comment by guardrex** [comment]
@rynowak Yes, all of the behaviors described in the OP are evidenced by the topic's sample app ...

https://github.com/aspnet/Docs/tree/master/aspnetcore/razor-pages/razor-pages-conventions/sample/

Make requests to ...

`/TheContactPage` and `/TheContactPage/somevalue`

Then, make a request for `/TheContactPage` (the app just loads the Index page) and pop open the "Other Pages" menu item and check the links there ... they're fouled.

**Comment by rynowak** [comment]
Alright dope, I'll take a look 👍

**Comment by rynowak** [comment]
Wow, that's a big repo 😆 - does it help if you comment out https://github.com/aspnet/Docs/blob/master/aspnetcore/razor-pages/razor-pages-conventions/sample/Conventions/GlobalTemplatePageRouteModelConvention.cs ?

I'm still downloading all of the code so I haven't been able to test it yet, but I suspect that convention is making it so that `/Index` will have a route template with order -1 of `/{globalTemplate?}`  - that would make `/TheContactPage` and since the order is -1 it will run before the route you added for `/TheContextPage/{text?}`.

**Comment by guardrex** [comment]
>  still downloading all of the code 

You hit something we hear about constantly from the community. Docs engineers are working on it ... something that will allow individual samples to be downloaded. https://github.com/aspnet/Docs/issues/6991#issuecomment-398104010

Yes, indeed on your early guess. The sample app is a bear trying to explain all of the scenarios. No, I didn't try commenting out the other scenarios. I'll try that now and let you know.
